### PR TITLE
Flash Attention for Neuron

### PR DESCRIPTION
**Pull request by apoorvtintin**
This PR adds support for flash attention kernel for Neuron implemented through Neuron Kernel Interface (NKI). 

The flash attention kernel works with TRN1 and TRN2.

### CODE REVIEW

File: axlearn/common/flash_attention/neuron_attention.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
+from absl import logging
+from functools import partial
+import jax
+import jax.numpy as jnp
+import jax.numpy as jnp
```
Comment (apivovarov) [inline axlearn/common/flash_attention/neuron_attention.py:5]: dup

File: axlearn/common/flash_attention/neuron_attention.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+lnc = 2 if jax.devices()[0].device_kind == "NC_v3d" else 1
+
+@partial(custom_vjp, nondiff_argnums=(4, 5))
+def flash_attention(query, key, value, bias, causal, softmax_scale):
+    out, _ = _mha_forward(query, key, value, bias, causal, softmax_scale)
+    return out
+
+
+def _mha_forward(query, key, value, bias, causal, softmax_scale):
```
Comment (kelvin-zou) [inline axlearn/common/flash_attention/neuron_attention.py:14]: Can we get a support for segment ID and dropout as well? Both are quite needed nowadays.

File: axlearn/common/flash_attention/utils.py [hunk 3]
```diff
@@ -159,6 +159,21 @@ def jit_attn(query, key, value, bias, segment_ids):
 
         return jit_attn
 
+    elif backend == "neuron":
+        from axlearn.common.flash_attention.neuron_attention import (
```
Comment (kelvin-zou) [inline axlearn/common/flash_attention/utils.py:163]: On demand import is kind of risky, we can live with it for functions inside neuron_attention.py, can we at least get it as a header import for files outside of neuron_attention.py?

File: axlearn/common/flash_attention/neuron_attention.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    # Get the batch size, sequence lengths, number of heads, and hidden dimension
+    batch_size, q_seq_len, num_heads, d_model = query.shape
+
+    # Transpose the query, key, and value tensors
+    q = query.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, q_seq_len]
+    k = key.transpose(0, 2, 3, 1)  # [batch_size, num_heads, d_model, kv_seq_len]
+    v = value.transpose(0, 2, 1, 3)  # [batch_size, num_heads, kv_seq_len, d_model]
+
+    import neuronxcc.nki.language as nl
```
Comment (kelvin-zou) [inline axlearn/common/flash_attention/neuron_attention.py:23]: Please add pylint here

**Review by ruomingp — COMMENTED** [review]
Will defer to @kelvin-zou for approval.

**Comment by apoorvtintin** [comment]
This has been implemented in a new PR https://github.com/apple/axlearn/pull/939. That PR addresses all comments in this PR but is from a different fork of Axlearn.

**Comment by apoorvtintin** [comment]
This has been implemented in a new PR https://github.com/apple/axlearn/pull/939. That PR addresses all comments in this PR but is from a different fork of Axlearn.

File: axlearn/common/flash_attention/neuron_attention.py (hunk 2 above)
Comment (apoorvtintin) [reply axlearn/common/flash_attention/neuron_attention.py:14]: Segment ID support is in progress and will be added soon
